### PR TITLE
Implement basic contentable functionality

### LINKED ISSUE #1: Main Contentable model needs to be implemented
## Description

There should be a main Polymorphic relation model that allows for any arbitrary model to be related to any other special "module" class / model.

This could also start as only being able to be related to a specific Page model, and then expanded to be relatable to all models.

### PR DESCRIPTION
**Pull request by m-triassi**
# Description
Closes #1 

Adds traits and interfaces that allow any model to be renderable or contentable. Adds base content pivot model. Adds helper functions to render set of modules to JSON or HTML in single function call.

### CODE REVIEW

File: src/Concerns/CanRender.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+trait CanRender
+{
+    public static function bootCanRender()
+    {
+        static::updated(function (Renderable $module) {
+            foreach ($module->renderable as $content) {
+                Contentable::clearCache($content->contentable->getKey());
+            }
+        });
```
Comment (kfriars) [inline src/Concerns/CanRender.php:17]: If you redefine the `renderable` relation to `content` (see comment below),  and also move the cache clearing logic to the Content Model/Interface (See other suggestion) this should become:

```suggestion
        static::updated(function (Renderable $renderable) {
            $renderable->contentable()?->clearContentCache();
        });
```

File: src/Concerns/CanRender.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        static::deleted(function (Renderable $module) {
+            foreach ($module->renderable as $content) {
+                Contentable::clearCache($content->contentable->getKey());
+            }
+        });
```
Comment (kfriars) [inline src/Concerns/CanRender.php:23]:
```suggestion
        static::deleted(function (Renderable $renderable) {
            $renderable->contentable()?->clearContentCache();
        });
```

File: src/Concerns/CanRender.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+        if (method_exists(new self(), 'bootSoftDeletes')) {
```
Comment (kfriars) [inline src/Concerns/CanRender.php:25]: The Framework and other packages tend to use:

```php
in_array(SoftDeletes::class, class_uses_recursive(static::class))
```

File: src/Concerns/CanRender.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            static::restored(function (Renderable $module) {
+                foreach ($module->renderable as $content) {
+                    Contentable::clearCache($content->contentable->getKey());
+                }
+            });
```
Comment (kfriars) [inline src/Concerns/CanRender.php:30]:
```suggestion
            static::restored(function (Renderable $renderable) {
                $renderable->contentable()?->clearContentCache();
            });
```

File: src/Contentable.php [hunk 5]
```diff
@@ -2,6 +2,18 @@
...
 {
+    public function clearCache($key)
+    {
+        if (Cache::has("contentable.html.{$key}")) {
+            Cache::delete("contentable.html.{$key}");
+        }
+
+        if (Cache::has("contentable.json.{$key}")) {
+            Cache::delete("contentable.json.{$key}");
+        }
```
Comment (kfriars) [inline src/Contentable.php:17]: I would move this to `HasContent` and add it to the `Contentable` interface as `public function clearCache(): void`. The "content cache" in the implementation is directly tied to the individual `Contentable` things.

File: src/Contracts/Contentable.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+use Illuminate\Database\Eloquent\Relations\MorphMany;
+
+/**
+ * @property string $render_order_column
+ */
+interface Contentable
+{
+    public function contents(): MorphMany;
+
+    public function scopeInRenderableOrder(Builder $q): void;
```
Comment (kfriars) [inline src/Contracts/Contentable.php:15]: Should be moved to the `Content` model imo.

File: src/Contracts/Contentable.php [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    public function attachContent(Renderable $renderable);
```
Comment (kfriars) [inline src/Contracts/Contentable.php:17]: From discussion above, I think this should be out of scope for this package.

**Review by kfriars — CHANGES_REQUESTED** [review]
Quite a bit of feedback here. I spent a good amount of time thinking this one through, and look forward to the further discussion of my comments.

File: src/Contentable.php (hunk 5 above)
Comment (m-triassi) [reply src/Contentable.php:17]: Yeah that's an excellent point. We don't really need this overarching class at all. Refactoring!

File: src/Concerns/HasContent.php [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    public function contents(): MorphMany
+    {
+        $contentModel = config('contentable.model');
+        return $this->morphMany($contentModel, 'contentable');
+    }
+
+    public function scopeInRenderableOrder(Builder $q): void
+    {
+        $q->orderBy($this->renderOrderColumnField());
+    }
```
[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/HasContent.php:24]: I think this might be a bygone artifact from a previous architecture, but i kept the ability to order things. I agree though, it makes sense for the order to come from the content model instead. I'll move that over

File: src/Concerns/CanRender.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+
+    public function renderable(): MorphMany
+    {
+        $contentModel = config('contentable.model');
+        return $this->morphMany($contentModel, 'renderable');
+    }
```
[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/CanRender.php:38]: This generally makes sense to me, but since the `content()` relation is a `MorphMany` it returns a collection. So i'm going to type the `contentable()` helper to optionally return a `Collection` instead. 

That also changes the logic you have above for the cache busting. Since we're working with collections, though i can just reach for a higher-order `->each->clearCache()`

File: src/Concerns/CanRender.php (hunk 3 above)
Comment (m-triassi) [reply src/Concerns/CanRender.php:25]: Fair enough, i definitely view that as less readable but i'm ok with sacrificing readability to be more inline with community standards

File: src/Concerns/CanRender.php (hunk 1 above)
Comment (m-triassi) [reply src/Concerns/CanRender.php:17]: As mentioned below, since we're getting back a `Collection` instead of a single instance, we need a call to `->each` (or a foreach lookp) here, but otherwise i agree with the suggestions!

File: src/Contracts/Contentable.php (hunk 6 above)
Comment (m-triassi) [reply src/Contracts/Contentable.php:15]: Agree'd!

File: src/Concerns/HasContent.php [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+            $renderables = collect($renderables);
+        }
+
+        return $renderables->map(function ($renderable) {
+            return [
+                'renderable_type' => $renderable::class,
+                'renderable_id' => $renderable->getKey()
+            ];
+        })->all();
+    }
```
[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/HasContent.php]: This piece of feedback is the only i'm kind of agonizing over the most. I'm not sure that if i'm fully convinced it is out of scope since we have a `MorphMany` relationship on `contents()` (for any `Contentable` model) the `sync()`method doesn't actually exist which makes it quite annoying to add new content. In the context of Nova, i'm almost certain this would cut down on the code needed in the resolvers.

If you feel really strongly though, we can drop these methods, and let the package consumer handle it

File: src/Contracts/Contentable.php (hunk 7 above)
Comment (m-triassi) [reply src/Contracts/Contentable.php:17]: i'll leave this unresolved, but we can have our discussion on this issue in the unresolved comment above

File: src/Concerns/HasContent.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (kfriars) [reply src/Concerns/HasContent.php]: I dont feel strongly about it. Let's leave it in, and if it feels like it wasn't adding much value, overcomplicating things or not providing the functionality we need, we can always take it out later and release a new major version.

If you want to keep it in, it should probably go to that one extra level where it can handle arbitrary data on the content model. Shouldn't be much from where you are. The `formatKeys` method should be more similar to `InteractsWithPivotTable@formatRecordsList`, to allow you to attach arbitrary data while creating the intermediate `Content` models.

ie: when you provide an array, you assume the user knew what they were doing and used yout `[$id => [...$attributes]]` format.

File: src/Concerns/CanRender.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (kfriars) [reply src/Concerns/CanRender.php]: This one I do have strong opinions about, but this package is your lead, so if you still feel strongly like $renderable -> $content should be a `hasMany` it is your call.

I think in our tradition Page/Modules approach we had it right. One Module is associated with one page. I don't see a case for a Renderable belonging to more than one Contentable. ie: a text module belonging to multiple pages. Further, inside of nova that wouldn even be possible without providing extra functionality for them to link their renderable onto some other contentable.


Used chatgpt for the following. Awesome:
```mermaid
classDiagram
    class Contentable {
    }
    class Content {
    }
    class Renderable {
    }
    Contentable "1" -- "0..*" Content : contents
    Content "1" -- "1" Renderable : renderable
```

File: src/Concerns/HasContent.php [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        foreach ($classes as $class => $ids) {
+            $contentModel::where('contentable_id', $this->getKey())
+                ->where('contentable_type', self::class)
+                ->where('renderable_type', $class)
+                ->whereIn('renderable_id', $ids)
+                ->delete();
+        }
+    }
+
+    private function formatKeys(Renderable|Collection|array $renderables)
```
Comment (kfriars) [inline src/Concerns/HasContent.php:111]: One small thing which can be ignored if you choose. But the typingting technically needs to change or the interface needs to add `getKey()`. I would opt for changing the typehint.

You also would need to import the Model class if you opt to make this change. I will add a suggestion above which you can choose to commit or ignore along with this.

```suggestion
    private function formatKeys((Model&Renderable)|Collection|array $renderables)
```

File: src/Concerns/HasContent.php [hunk 12]
```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace Plank\Contentable\Concerns;
+
+use Illuminate\Database\Eloquent\Collection as EloquentCollection;
```
Comment (kfriars) [inline src/Concerns/HasContent.php:5]: Paired with other comment.

```suggestion
use Illuminate\Database\Eloquent\Collection as EloquentCollection;
use Illuminate\Database\Eloquent\Model;
```

File: src/Concerns/CanRender.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/CanRender.php]: Ah i see, i assumed you wanted to keep it `MorphMany` based off your code suggestion. I think theres some good logic behind making it `MorphOne` i can look into doing that

[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/CanRender.php]: I made this change and it seems to work just fine with all the existing code! Lets stick with it :)

File: src/Concerns/HasContent.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/HasContent.php]: I like that plan, let's check back in later with these functions and evaluate if they end up being useful. 

Im not sure i follow what you mean in the second paragraph though -- currently the sync / attach / detach function allow you to pass in any renderable content (array or otherwise), and attaches them while creating any needed intermediary content entries. This should be working exactly the same way `sync()` works on the vanilla many-to-* relations; pass data, get pivots created, report changes.

Maybe i'm misinterpreting what you mean? lmk!

**Comment by m-triassi** [comment]
Ok (i think) we're almost ready for prime time? re-requesting review 🤞

File: src/Concerns/HasContent.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (kfriars) [reply src/Concerns/HasContent.php]: I don't see it in the implementation. I will test it out and report back. Maybe my 🧠 is just 💤 ...

[the remark this answers is not in the document]
Comment (kfriars) [reply src/Concerns/HasContent.php]: Woops. Meant to reply here earlier. The following is why I said this feels "out of scope".

As far as I can see and have tested, this implementation does not support atttaching arbitrary data to the Content pivot. To do it well "in my opinioin", would involve adding substantial amounts of logic and creating and enforcing a new structure to the `sync`, `attach`, etc. methods than Laravel users are used to and expect.

Here is a sample test I created, that still has a glaring issue but highlights the functionality I am referencing:

```php
it('allows you to add custom data to each pivot when attaching many renderables using attachContent', function () {
    expect(Content::query()->count())->toEqual(0);

    $page = Page::factory()->create();
    [$a, $b, $c] = FakeModule::factory()->count(3)->create()->toArray();

    $page->attachContent([
        $a->id => [
            'identifier' => 'First',
            'order' => 1,
            'narration' => 'I want this one to go first!'
        ],
        $b->id => [
            'identifier' => 'Second',
            'order' => 2,
            'narration' => 'I want this one to go second!'
        ],
        $c->id => [
            'identifier' => 'Third',
            'order' => 3,
            'narration' => 'I want this one to go third!'
        ]
    ]);

    [$one, $two, $three] = Content::query()->orderByDesc('order')->get()->toArray();
    
    expect($one->contentable_id)->toEqual($page->id);
    expect($one->contentable_type)->toEqual($page::class);
    expect($one->renderable_id)->toEqual($a->id);
    expect($one->renderable_type)->toEqual($a::class);
    expect($one->order)->toEqual(1);
    expect($one->narration)->toEqual('I want this one to go first!');

    expect($two->contentable_id)->toEqual($page->id);
    expect($two->contentable_type)->toEqual($page::class);
    expect($two->renderable_id)->toEqual($b->id);
    expect($two->renderable_type)->toEqual($b::class);
    expect($two->order)->toEqual(2);
    expect($two->narration)->toEqual('I want this one to go second!');

    expect($three->contentable_id)->toEqual($page->id);
    expect($three->contentable_type)->toEqual($page::class);
    expect($three->renderable_id)->toEqual($c->id);
    expect($three->renderable_type)->toEqual($c::class);
    expect($three->order)->toEqual(3);
    expect($three->narration)->toEqual('I want this one to go third!');
});
```

Here we are attaching arbitrary pivot data to the Content model, and I think it is reasonable to expect package consumers will have their own needs on their Content models.

The above format is what Laravel users are used to for syncing/attaching many records to pivots. The glaring issue is that this is not sufficient for our polymorph to polymorph pivot table. The reason being, the `$id` defined in these arrays is not truly identifying, since it is missing its accompanying `renderable_type`.

You would need to create a data structure and convention, and rewrite mostly allof the logic of `InteractsWithPivotTable` to provide the `attach`, `sync` etc. methods you are trying to recreate here.

This is illegal syntax, but demonstrates the new type of data structure I think would make this possible:
```php
    $page->attachContent([
        [$a->id, $a::class] => [ ...$attributes],
    ]);
```
You could add a convention like `[$renderable, [...$attributes]]` but you would still need to rewrite all of the logic I referenced earlier to accommodate this.

Your current `syncContent` method provides no opportunity for any pivot data to be created/update.
Your `attachContent` method only supports an `identifier` when passed a single `Renderable`, or attaching multiple `Renderable`s with the same identifier.

******

So in summary, my original opinion that this is out of scope for the package holds. If choosing between working directly with the `Content` model in a way you are already familiar with as a Laravel user, versus writing a good chunk of code + forcing package consumers to learn your new convention – for now at least – I would lean towards the former.

[the remark this answers is not in the document]
Comment (m-triassi) [reply src/Concerns/HasContent.php]: Hmm i see what you mean now, yeah you're right there isn't a clean way to pass pivot data in the function 🤔 That's really unfortunate i would have really preferred to keep this in the package. Oh well, i'll rip it out then!

[the remark this answers is not in the document]
Comment (kfriars) [reply src/Concerns/HasContent.php]: If you were to implement this functionality, I honestly think it could warrant its own package.

A 'morphAny' pseudo relationship of sorts.

File: src/Contracts/Content.php [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+interface Content
+{
+    public function renderable(): MorphTo;
+
+    public function contentable(): MorphTo;
+
+    public function scopeInRenderableOrder(Builder $query): void;
+
+    public function renderOrderColumnField(): string;
```
Comment (kfriars) [inline src/Contracts/Content.php:16]: I think it is good to provide this in the default Model we ship, since I assume it is a 90%+ usecase, but I think people should be able to ship Content models without ordering.

You don't consume these methods from the interface anywhere else in the package, so should be safe to remove. If you rely on this elsewhere in the package in the future, we can add them into the contract then.

```suggestion
```

**Review by kfriars — APPROVED** [review]
Looks good. A couple thoughts, but can totally 🚢 as is.

File: src/Contracts/Content.php (hunk 13 above)
Comment (m-triassi) [reply src/Contracts/Content.php:16]: fair enough!
